testinglogger/testinglogger.go

```diff
@@ -0,0 +1,134 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+Copyright 2020 Intel Coporation.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package testinglogger contains an implementation of the logr interface
+// which is logging through a function like testing.TB.Log function.
+// Therefore it can be used in standard Go tests and Gingko test suites
+// to ensure that output is associated with the currently running test.
+//
+// Serialization of the structured log parameters is done in the same way
+// as for klog.InfoS.
+//
+package testinglogger
+
+import (
+	"bytes"
+
+	"github.com/go-logr/logr"
+
+	"k8s.io/klog/v2/internal/serialize"
+)
+
+// Logger is the relevant subset of testing.TB.
+type Logger interface {
+	Helper()
+	Log(args ...interface{})
+}
+
+// New constructs a new logger for the given test interface.
+func New(l Logger) logr.Logger {
```

Comment:
Is this significantly different than logr/testing.NewTestLogger() ?  I didn't have time to read all of this yet.

Reply:
The output formatting is done the same way as in klog. I don't know what the funcr formatting looks like, so I can't say how different it is. One of my goals was to have the same output in tests as in normal Kubernetes logs. If this isn't considered important enough to have a second logger that wraps `testing.T`, then we can also drop this PR.

But `logr/testing.NewTestLogger` has the same problem with incorrect stack output as this code here:
```
/nvme/gopath/src/github.com/go-logr/logr/testing$ go test -v .
=== RUN   TestTestLogger
    test.go:30: : "level"=0 "msg"="info"
    test.go:30: : "level"=0 "msg"="V(0).info"
    test.go:30: : "msg"="error" "error"="error"
--- PASS: TestTestLogger (0.00s)
PASS
```

Reply:
I can see the problem with attribution when I pay attention to logr/testing.NewTestLogger()  (which itself needs to call t.Helper()).  So I guess I endorse *a* solution if we can figure out the least painful one

Reply:
and at that point, hopefully most of this impl can be replaced by logr/testing.NewTestLogger()

Reply:
> I can see the problem with attribution when I pay attention to logr/testing.NewTestLogger() (which itself needs to call t.Helper()).

That just shifts the attribution up one level into funcr.go. We need to skip two additional levels (funcr.go and logr.go) and can't do that with the current T API.

```
/nvme/gopath/src/github.com/go-logr/logr/testing$ git diff
diff --git a/testing/test.go b/testing/test.go
index 98f7f78..c95ea21 100644
--- a/testing/test.go
+++ b/testing/test.go
@@ -27,6 +27,7 @@ import (
 // Info logs are only enabled at V(0).
 func NewTestLogger(t *testing.T) logr.Logger {
        fn := func(prefix, args string) {
+               t.Helper()
                t.Logf("%s: %s", prefix, args)
        }
        return funcr.New(fn, funcr.Options{})
/nvme/gopath/src/github.com/go-logr/logr/testing$ go test -v .
=== RUN   TestTestLogger
    funcr.go:274: : "level"=0 "msg"="info"
    funcr.go:274: : "level"=0 "msg"="V(0).info"
    funcr.go:291: : "msg"="error" "error"="error"
--- PASS: TestTestLogger (0.00s)
PASS
ok  	github.com/go-logr/logr/testing	(cached)
```